src/chat_commands.c

```diff
@@ -335,6 +337,117 @@ void cmd_game_join(WINDOW *window, ToxWindow *self, Tox *tox, int argc, char (*a
 
 #endif // GAMES
 
+
+void cmd_fopen(WINDOW *window, ToxWindow *self, Tox *tox, int argc, char (*argv)[MAX_STR_SIZE])
+{
+    UNUSED_VAR(window);
+
+    if (argc < 1) {
+        line_info_add(self, false, NULL, NULL, SYS_MSG, 0, 0, "File ID required.");
+        return;
+    }
+
+    long int idx = strtol(argv[1], NULL, 10);
+
+    if ((idx == 0 && strcmp(argv[1], "0")) || idx < 0 || idx >= MAX_FILES) {
+        line_info_add(self, false, NULL, NULL, SYS_MSG, 0, 0, "No pending file transfers with ID %ld", idx);
+        return;
+    }
+
+    FileTransfer *ft = get_file_transfer_struct_index(self->num, idx, FILE_TRANSFER_RECV);
+
+    if (!ft) {
+        line_info_add(self, false, NULL, NULL, SYS_MSG, 0, 0, "No pending file transfers with ID %ld.", idx);
+        return;
+    }
+
+    if (ft->state != FILE_TRANSFER_PENDING) {
+        line_info_add(self, false, NULL, NULL, SYS_MSG, 0, 0, "No pending file transfers with ID %ld.", idx);
+        return;
+    }
+
+    if ((ft->file = fopen(ft->file_path, "a")) == NULL) {
+        const char *msg =  "File transfer failed: Invalid download path.";
+        close_file_transfer(self, tox, ft, TOX_FILE_CONTROL_CANCEL, msg, notif_error);
+        return;
+    }
+
+    Tox_Err_File_Control err;
+    tox_file_control(tox, self->num, ft->filenumber, TOX_FILE_CONTROL_RESUME, &err);
+
+    if (err != TOX_ERR_FILE_CONTROL_OK) {
+        goto on_recv_error;
+    }
+
+    // make tmp_dir if it does not exist
+    /// don't need this for now, prob should go somewhere else anyway.
+    const char *tmp_dir = "/tmp/toxic-download-dir/";
+    DIR* dir = opendir("/tmp/toxic-download-dir/");
+    if (dir) {
+        closedir(dir);
+    }
+
+    if (mkdir(tmp_dir, S_IRWXU) == -1) {
+        line_info_add(self, false, NULL, NULL, SYS_MSG, 0, 0, "Could not create tox download /tmp/ directory.\n", err);
+    }
+    // end make tmpdir if it does not exist
+
+    line_info_add(self, false, NULL, NULL, SYS_MSG, 0, 0, "The file path is '%s'", ft->file_path);
+
+    line_info_add(self, false, NULL, NULL, SYS_MSG, 0, 0, "Saving file [%ld] as: '%s'", idx, ft->file_path);
+
+    const bool auto_accept_files = friend_get_auto_accept_files(self->num);
+    const uint32_t line_skip = auto_accept_files ? 4 : 2;
+
+    char progline[MAX_STR_SIZE];
+    init_progress_bar(progline);
+    line_info_add(self, false, NULL, NULL, SYS_MSG, 0, 0, "%s", progline);
+    ft->line_id = self->chatwin->hst->line_end->id + line_skip;
+    ft->state = FILE_TRANSFER_STARTED;
+
+    // make and call xdg command
+    while (ft->state != FILE_TRANSFER_INACTIVE) {
+        usleep(100000);
+    }
```

Comment:
there are legitimate uses for opening files that are still being transferred (most streaming media types like videos and audio)

Reply:
That sleep will freeze toxic. ft->state can't update while we're in that loop because that function is holding a mutex lock. Sleeps in random functions is almost never the correct solution. They're generally only used in the main loop or in task-specific threads.